sql/tests: TestDependencyDigestOptimization failed

sql/tests.TestDependencyDigestOptimization [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18521860?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18521860?buildTab=artifacts#/) on master @ [9434611a624498c38878da8bc94eea585d66cee6](https://github.com/cockroachdb/cockroach/commits/9434611a624498c38878da8bc94eea585d66cee6):


```
=== RUN   TestDependencyDigestOptimization
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/0bb9b7b43cd490b3d5c954f3b4506b18/logTestDependencyDigestOptimization2310189798
    test_log_scope.go:76: use -show-logs to present logs inline
    test_server_shim.go:152: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
    dependency_digest_test.go:140: 
        	Error Trace:	pkg/sql/tests/dependency_digest_test.go:140
        	Error:      	Should be true
        	Test:       	TestDependencyDigestOptimization
        	Messages:   	did not find re-optimized plan
    panic.go:626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/0bb9b7b43cd490b3d5c954f3b4506b18/logTestDependencyDigestOptimization2310189798
I250116 10:23:44.014737 1281 kv/kvserver/queue.go:1341  [T1,Vsystem,n1,lease] 2  purgatory is now empty
I250116 10:23:44.030268 1193 2@util/log/event_log.go:39  [T1,Vsystem,n1] 3 ={"Timestamp":1737023024030262789,"EventType":"runtime_stats","MemRSSBytes":638046208,"GoroutineCount":3291,"MemStackSysBytes":38338560,"GoAllocBytes":351199864,"GoTotalBytes":460549480,"HeapFragmentBytes":43318664,"HeapReservedBytes":6299648,"HeapReleasedBytes":5193728,"CGoAllocBytes":306469312,"CGoTotalBytes":501612544,"CGoCallRate":0.00007519647,"GCPausePercent":0.0000000032492853,"GCRunCount":19,"NetHostRecvBytes":100329124,"NetHostSendBytes":100329124}
I250116 10:23:44.030424 1193 2@server/status/runtime_log.go:43  [T1,Vsystem,n1] 4  runtime stats: 608 MiB RSS, 3291 goroutines (stacks: 37 MiB), 335 MiB/439 MiB Go alloc/total (heap fragmentation: 41 MiB, heap reserved: 6.0 MiB, heap released: 5.0 MiB), 292 MiB/478 MiB CGO alloc/total (0.0 CGO/sec), 0.0/0.0 %(u/s)time, 0.0 %gc (19x), 96 MiB/96 MiB (r/w)net
--- FAIL: TestDependencyDigestOptimization (10.28s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDependencyDigestOptimization.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-46558

sql/tests.TestDependencyDigestOptimization [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18529624?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18529624?buildTab=artifacts#/) on master @ [f118d515eb13bf5647b16121900278fb6e191183](https://github.com/cockroachdb/cockroach/commits/f118d515eb13bf5647b16121900278fb6e191183):


```
=== RUN   TestDependencyDigestOptimization
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/0bb9b7b43cd490b3d5c954f3b4506b18/logTestDependencyDigestOptimization2328385043
    test_log_scope.go:76: use -show-logs to present logs inline
I250116 22:50:33.559409 1 (gostd) rand.go:256  [-] 1  random seed: -1054279678284117939
    dependency_digest_test.go:140: 
        	Error Trace:	pkg/sql/tests/dependency_digest_test.go:140
        	Error:      	Should be true
        	Test:       	TestDependencyDigestOptimization
        	Messages:   	did not find re-optimized plan
    panic.go:626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/0bb9b7b43cd490b3d5c954f3b4506b18/logTestDependencyDigestOptimization2328385043
--- FAIL: TestDependencyDigestOptimization (11.74s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDependencyDigestOptimization.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/sql/tests/dependency_digest_test.go
+++ b/pkg/sql/tests/dependency_digest_test.go
@@ -10,10 +10,12 @@ import (
 	"testing"
 
 	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/errors"
 	"github.com/stretchr/testify/require"
 )
 
@@ -128,16 +130,24 @@ CREATE FUNCTION db1.sc2.f1 (input INT) RETURNS INT8
 	}
 	require.Truef(t, foundReusedPlan, "did not find reused plan")
 	changeConn.Exec(t, "CREATE STATISTICS s1 FROM db2.sc3.t1")
-	explainRows = prepareConn.QueryStr(t, "EXPLAIN ANALYZE EXECUTE p1")
-	foundReOptimizedPlan := false
-	for _, row := range explainRows {
-		// New stats were generated so the plan should get re-optimized.
-		if row[0] == "plan type: generic, re-optimized" {
-			foundReOptimizedPlan = true
-			break
+	// Statistics cache is updated via a range feed, so propagation may take
+	// time.
+	testutils.SucceedsSoon(t, func() error {
+		explainRows = prepareConn.QueryStr(t, "EXPLAIN ANALYZE EXECUTE p1")
+		foundReOptimizedPlan := false
+		for _, row := range explainRows {
+			// New stats were generated so the plan should get re-optimized.
+			if row[0] == "plan type: generic, re-optimized" {
+				foundReOptimizedPlan = true
+				break
+			}
 		}
-	}
-	require.Truef(t, foundReOptimizedPlan, "did not find re-optimized plan")
+		if !foundReOptimizedPlan {
+			return errors.AssertionFailedf("did not find re-optimized plan")
+		}
+		return nil
+	})
+
 	// Confirm changing the role will have trouble.
 	prepareConn.Exec(t, "SET ROLE bob")
 	row := query.QueryRow()

